src/clients/UBAClient/UBAClientUtilities.ts

```diff
@@ -77,3 +127,336 @@ export async function getUBAFeeConfig(
+        chainId,
+        tokenSymbol,
+        hubPoolClient,
+        hubPoolBlockNumber
+      );
+      // Find the bundle that has the same block number as the current block number
+      const referenceBundleIndex = availableBundles.findLastIndex((bundle) => bundle.blockNumber === blockNumber);
+
+      const tokenMappingLookup = (
+        TOKEN_SYMBOLS_MAP as Record<string, { addresses: { [x: number]: string }; decimals: number }>
+      )[tokenSymbol];
+      const hubPoolTokenAddress = tokenMappingLookup.addresses[hubPoolClient.chainId];
+      const tokenDecimals = tokenMappingLookup.decimals;
+      const erc20 = ERC20__factory.connect(hubPoolTokenAddress, hubPoolClient.hubPool.provider);
+      const [hubBalance, hubEquity, ethSpokeBalance, spokeTargets] = await Promise.all([
+        erc20.balanceOf(hubPoolClient.hubPool.address, { blockTag: blockNumber }),
+        erc20.balanceOf(hubPoolClient.hubPool.address, { blockTag: blockNumber }),
+        erc20.balanceOf(spokePoolClient.spokePool.address, { blockTag: blockNumber }),
+        hubPoolClient.configStoreClient.getUBATargetSpokeBalances([chainId], hubPoolTokenAddress, blockNumber),
+      ]);
+
+      // Construct the bundle. If the bundle already exists, use the existing bundle
+      const constructedBundle: UBABundleState = {
+        ...(referenceBundleIndex !== -1 ? availableBundles[referenceBundleIndex] : { flows: [] }),
+        blockNumber,
+        openingBalance: spokePoolBalance,
+        openingIncentiveBalance: incentiveBalance,
+        config: {
+          ubaConfig: await getUBAFeeConfig(chainId, tokenSymbol, blockNumber),
+          tokenDecimals,
+          hubBalance,
+          hubEquity,
+          hubPoolSpokeBalance: ethSpokeBalance,
+          spokeTargets,
+        },
+      };
+      // If the bundle already exists, replace it
+      if (referenceBundleIndex !== -1) {
+        availableBundles[referenceBundleIndex] = constructedBundle;
+      }
+      // Otherwise, add it to the list of bundles
+      else {
+        availableBundles.push(constructedBundle);
+      }
+      // Find the last flow avaialble in the bundle
+      const lastFlow = constructedBundle.flows[constructedBundle.flows.length - 1];
+      // If the last flow exists, call getFlows with the last flow's block number as the fromBlock
+      const recentFlows = getFlows(
+        chainId,
+        relevantChainIds,
+        spokePoolClients,
+        hubPoolClient,
+        lastFlow?.flow.blockNumber ?? chainState.spokeChain.bundleEndBlockNumber
+      );
+      // Instantiate a UBAFeeSpokeCalculator
+      const calculator = new UBAFeeSpokeCalculator(
+        chainId,
+        tokenSymbol,
+        [...constructedBundle.flows.map(({ flow }) => flow), ...recentFlows],
+        spokePoolBalance,
+        incentiveBalance,
+        constructedBundle.config.ubaConfig
+      );
+      for (const flow of recentFlows) {
```

Comment:
nit: I assume `recentFlows` is returned in ascending order, right? Just want to make sure you've seen this helpful functions to deterministically sort events on the same chain: https://github.com/across-protocol/relayer-v2/blob/master/src/utils/EventUtils.ts#L181

Reply:
Yep! The end of `getFlows()` returns the flows in ascending order.

Reply:
Added docs to make this more explicit: https://github.com/across-protocol/sdk-v2/pull/266/commits/a03635d2b538c3ebc003b2e240857b44d4d17b0f